routes/merged/project/vm.rb

```diff
@@ -2,14 +2,20 @@
 
 class Clover
   branch = lambda do |r|
-    vm_endpoint_helper = Routes::Common::VmHelper.new(app: self, request: r, user: current_account, location: nil, resource: nil)
-
     r.get true do
-      vm_endpoint_helper.list
+      dataset = vm_list_dataset
+
+      if api?
+        vm_list_api_response(dataset)
+      else
+        @vms = Serializers::Vm.serialize(dataset.eager(:semaphores, :assigned_vm_address, :vm_storage_volumes).reverse(:created_at).all, {include_path: true})
```

Comment:
Huh, why is the eager load of semaphores interesting? Incidentally, loading of correlated objects (eager or not) is where we might get into a bit of trouble, e.g. if vm_storage_volumes has its own "view" capability or lack thereof. 

This is kind of the crux of the sensitivity we have to authorization bypasses for something as ABAC with a policy language. Often, dealing with correlated objects is implicated.

The AWS and other consoles tend to break (as in, spew errors sometimes) because they can't load certain data given the permutations of what their policy languages allow for, should one make a minimal set of capabilities is made available to a principal.

If I had to choose one, I'd still choose a flexible policy language and a weird looking or slightly busted console, hence, Ubicloud is rather similar in that regard. But, making you aware that this is where the hairy stuff comes from.

Reply:
> Huh, why is the eager load of semaphores interesting

We use semaphores in the `display_state` https://github.com/ubicloud/ubicloud/blob/027c2c7d17d0ad266622d05bd612f9022b6defc1/model/vm.rb#L61